File: 3264-final-array-state-after-k-multiplication-operations-i/3264-final-array-state-after-k-multiplication-operations-i.py
from sortedcontainers import SortedList

class Solution:
    def getFinalState(self, nums: list[int], k: int, multiplier: int) -> list[int]:
        st = SortedList((val, idx) for idx, val in enumerate(nums))
        
        for _ in range(k):
            val, idx = st.pop(0)
            nums[idx] *= multiplier
            st.add((nums[idx], idx))
        
        return nums

File: 3264-final-array-state-after-k-multiplication-operations-i/3264-final-array-state-after-k-multiplication-operations-i.js
class MinHeap {
    constructor() {
        this.heap = [];
    }

    push([value, index]) {
        this.heap.push([value, index]);
        this.heap.sort((a, b) => a[0] - b[0] || a[1] - b[1]); 
    }

    pop() {
        return this.heap.shift(); 
    }

    size() {
        return this.heap.length;
    }
}

var getFinalState = function(nums, k, multiplier) {
    const heap = new MinHeap();

    nums.forEach((num, idx) => heap.push([num, idx]));

    while (k--) {
        const [value, idx] = heap.pop();
        nums[idx] *= multiplier;
        heap.push([nums[idx], idx]);
    }

    return nums;
};
